### PR TITLE
[flutter_adaptive_scaffold] Add breakpoint extension to get the current active breakpoint

### PR DESCRIPTION
**Pull request by martijn00**
This enables people to check what the current breakpoint is and adjust UI based on that.

*List which issues are fixed by this PR. You must list at least one issue.*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

**Comment by martijn00** [comment]
@gspencergoog I just wanted to get your opinion on this before working further on it.

It needs a rebase after #7300 is merged

File: packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart
Comment (gspencergoog) [inline packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart:177]: So, if I set my own custom breakpoint, `activeBreakpoint` won't return it, right? And in fact, it will return one that is different from the actual value.  It think it would be better to return null here, and rename it to something that indicates that it's one of the standard ones (e.g. activeStandardBreakpoint) or something. And while this is still useful, it's more complicated to understand what it does.

Ideally, you'd find out the actual active breakpoint somehow and return that rather than guessing that it's `Breakpoints.standard`.  That would be easy to use and understand.

Comment (martijn00) [reply packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart:177]: Well you can only set actual breakpoints when using a SlotLayout somewhere. So this looks that up, but if you don't use a SlotLayout still returns the currently active config.

**Comment by martijn00** [comment]
@gspencergoog @Renzo-Olivares do you have some more feedback on this? Otherwise i'll go ahead and add some tests and other changes required.

File: packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart
Comment (Renzo-Olivares) [inline packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart:137]: Our style guide is partial against `extension` methods https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#avoid-using-extension . Maybe we can accomplish this through a static method in the `Breakpoint` class?

```dart
  static Breakpoint activeBreakpointOf(BuildContext context) {
    final SlotLayout? slotLayout = context.findAncestorWidgetOfExactType<SlotLayout>();
    if (slotLayout != null) {
      for (final MapEntry<Breakpoint, SlotLayoutConfig?> config
          in slotLayout.config.entries) {
        if (config.key.isActive(context)) {
          return config.key;
        }
      }
    }

    final TargetPlatform platform = Theme.of(context).platform;

    if (Breakpoint.desktop.contains(platform)) {
      if (Breakpoints.extraLarge.isActive(this)) {
        return Breakpoints.extraLargeDesktop;
      } else if (Breakpoints.large.isActive(this)) {
        return Breakpoints.largeDesktop;
      } else if (Breakpoints.mediumLarge.isActive(this)) {
        return Breakpoints.mediumLargeDesktop;
      } else if (Breakpoints.medium.isActive(this)) {
        return Breakpoints.mediumDesktop;
      } else if (Breakpoints.small.isActive(this)) {
        return Breakpoints.smallDesktop;
      }
    } else if (Breakpoint.mobile.contains(platform)) {
      if (Breakpoints.extraLarge.isActive(this)) {
        return Breakpoints.extraLargeMobile;
      } else if (Breakpoints.large.isActive(this)) {
        return Breakpoints.largeMobile;
      } else if (Breakpoints.mediumLarge.isActive(this)) {
        return Breakpoints.mediumLargeMobile;
      } else if (Breakpoints.medium.isActive(this)) {
        return Breakpoints.mediumMobile;
      } else if (Breakpoints.small.isActive(this)) {
        return Breakpoints.smallMobile;
      }
    }
    return Breakpoints.standard;
  }
```

**Comment by martijn00** [comment]
@Renzo-Olivares @gspencergoog this is ready to review now.

**Review by gspencergoog — APPROVED** [review]
32384589-a60f0e74-c078-11e7-9bc1-e5b5287aea9d

File: packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart
Comment (diegotori) [inline packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart:242]: Consider spinning this off into a separate method that maybe returns the breakpoint from `SlotLayout` if available:

```dart
static Breakpoint maybeGetBreakpointFromSlotLayout(BuildContext context) {
    final SlotLayout? slotLayout =
        context.findAncestorWidgetOfExactType<SlotLayout>();
    Breakpoint? fallbackBreakpoint;
    if (slotLayout != null) {
      for (final MapEntry<Breakpoint, SlotLayoutConfig?> config
          in slotLayout.config.entries) {
        if (config.key.isActive(context)) {
          if (config.key.platform != null) {
            return config.key;
          } else {
            fallbackBreakpoint ??= config.key;
          }
        }
      }
    }
    return fallbackBreakpoint;
}

```

File: packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart
Comment (diegotori) [inline packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart:278]: Consider also breaking this down into a smaller method, which gets the breakpoint based on current plaform:

```dart
static Breakpoint getBreakpointForTargetPlatform(BuildContext context) {
  final TargetPlatform platform = Theme.of(context).platform;
  final bool isDesktop = Breakpoint.desktop.contains(platform);
  final bool isMobile = Breakpoint.mobile.contains(platform);
  
  for (final Breakpoint breakpoint in <Breakpoint>[
    Breakpoints.small,
    Breakpoints.medium,
    Breakpoints.mediumLarge,
    Breakpoints.large,
    Breakpoints.extraLarge,
  ]) {
    if (breakpoint.isActive(context)) {
      if (isDesktop) {
        switch (breakpoint) {
          case Breakpoints.small:
            return Breakpoints.smallDesktop;
          case Breakpoints.medium:
            return Breakpoints.mediumDesktop;
          case Breakpoints.mediumLarge:
            return Breakpoints.mediumLargeDesktop;
          case Breakpoints.large:
            return Breakpoints.largeDesktop;
          case Breakpoints.extraLarge:
            return Breakpoints.extraLargeDesktop;
          default:
            return Breakpoints.standard;
        }
      } else if (isMobile) {
        switch (breakpoint) {
          case Breakpoints.small:
            return Breakpoints.smallMobile;
          case Breakpoints.medium:
            return Breakpoints.mediumMobile;
          case Breakpoints.mediumLarge:
            return Breakpoints.mediumLargeMobile;
          case Breakpoints.large:
            return Breakpoints.largeMobile;
          case Breakpoints.extraLarge:
            return Breakpoints.extraLargeMobile;
          default:
            return Breakpoints.standard;
        }
      } else {
        return breakpoint;
      }
    }
  }
  return Breakpoints.standard;
}
```

**Comment by diegotori** [comment]
@gspencergoog I know that you have already approved it, but I did add some feedback that might make @martijn00's changes a bit more concise.

In other words, I added suggestions in terms of breaking up the static method into smaller components, thus making it more manageable.

Please let me know if it makes sense for you to rescind your approval for these proposed suggestions. Thanks.

**Comment by martijn00** [comment]
@diegotori some good feedback! I've made the changes for it. @gspencergoog can you review again and let me know if you agree?

File: packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart
Comment (diegotori) [inline packages/flutter_adaptive_scaffold/lib/src/breakpoints.dart:302]: Simplify this to:

```dart
static Breakpoint activeBreakpointOf(BuildContext context) {
  return maybeActiveBreakpointFromSlotLayout(context) ?? defaultBreakpointOf(context);
}
```

**Comment by diegotori** [comment]
> @diegotori some good feedback! I've made the changes for it. @gspencergoog can you review again and let me know if you agree?

Added one more nit.

**Review by gspencergoog — APPROVED** [review]
Works for me.

File: packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart
Comment (martijn00) [inline packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart:26]: @diegotori this pointed a bug out to me and i'm struggling to find a solution right now. This is why the test fails.

The problem is Breakpoints.standard is always added, in the past the was ignored because a null value was returned from an earlier breakpoint. Because we check now for the best match instead of the first it returns the standard as well...

How can we get a reliable find for the best match?

Comment (diegotori) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart:26]: @martijn00 unfortunately, I don't have a lot of insight into this library in terms of how it selects the correct breakpoint.

@gspencergoog any thoughts on what @martijn00 posted?

**Review by diegotori — COMMENTED** [review]
LGTM.

File: packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart
Comment (diegotori) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart:26]: @martijn00 I did a bit of poking, and I see where you're getting caught.

Your test is failing because you're expecting no widget to be found, when in reality, one exists in `adaptive_scaffold_demo.dart`. The reason being is that it declares a "default" `secondaryBody` that will always exist at the `medium` Breakpoint and any time under `Breakpoints.standard`. Your `smallSecondaryBody` doesn't override that config due to it being set to `AdaptiveScaffold.emptyBuilder`.

As a result, when you run your tests, there will always be a `secondaryBody` because one was declared in the widget under test. There will be no `smallSBody` obviously because you set it to `AdaptiveScaffold.emptyBuilder` in the widget under test.

TL;DR: change the failing `sBody` assertion to `findsOneWidget` instead of `findsNothing`.

@gspencergoog please let me know if this is the expected behavior, because it looks to me like it is. Compared to the way body is set up, the reason those tests pass is because when you set `smallBody` to something other than `AdaptiveScaffold.emptyBuilder`, it will display that instead of `body`. If that value was set to `AdaptiveScaffold.emptyBuilder`, then `body` will send back the widget for it.

Comment (diegotori) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart:26]: @martijn00 Looks like I spoke too soon. I see the bug that you pointed out. I think I have a possible fix for this. Long story short, it was retaining your fallback widget if the current picked widget value is `null`. In other words, consider the following code:

```dart
        if (breakpoint.platform != null) {
          chosenWidget = pickedWidget;
        } else if (chosenWidget == null && pickedWidget != null) {
          fallbackWidget = pickedWidget;
        }
```

If the next picked widget was `null`, it would still hold on to the one that did apply previously. In your case, if you for example set `smallSecondaryBody` to `AdaptiveScaffold.emptyBuilder`, that should clear out the current secondary widget for that slot. However, this code doesn't account for null widgets, where it should reset that value if you run into a null picked widget. When modifying it to the following:

```dart
        if (breakpoint.platform != null) {
          chosenWidget = pickedWidget;
        } else if (chosenWidget == null && pickedWidget != null) {
          fallbackWidget = pickedWidget;
        } else if (fallbackWidget != null && pickedWidget == null) {
          // We previously had a fallback widget, but our slot has no widget.
          // That means don't display anything for it.
          fallbackWidget = null;
        }
```

It will actually return no widget for `AdaptiveScaffold.smallSecondaryBody` in `adative_scaffold_demo.dart`. You can also do likewise for the other secondary body properties, and it will work as intended: not displaying anything for a given slot when set to `AdaptiveScaffold.emptyBuilder`.

As a result, it passes the current tests as you wrote them.

Comment (diegotori) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart:26]: @martijn00 when I ran that method through Gemini, I was able to optimize it further:

```dart
  /// Given a context and a config, it returns the [SlotLayoutConfig] that will be
  /// chosen from the config under the context's conditions.
  static SlotLayoutConfig? pickWidget(
      BuildContext context, Map<Breakpoint, SlotLayoutConfig?> config) {
    SlotLayoutConfig? chosenWidget;

    // Iterate through breakpoints (assuming smallest to largest)
    for (final Breakpoint breakpoint in config.keys) {
      if (breakpoint.isActive(context)) {
        final SlotLayoutConfig? pickedWidget = config[breakpoint];
        if (pickedWidget != null) {
          if (breakpoint.platform != null) {
            return pickedWidget; // Prioritize platform-specific breakpoints
          } else {
            chosenWidget = pickedWidget; // Fallback to non-platform-specific
          }
        } else {
          chosenWidget = null; // Clear fallback if current pickedWidget is null
        }
      }
    }

    return chosenWidget;
  }
```

Comment (martijn00) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart]: @diegotori thanks! That does the trick!

**Comment by diegotori** [comment]
@gspencergoog looks like the PR's all green.

**Comment by martijn00** [comment]
@Renzo-Olivares @gspencergoog What do you think?

File: packages/flutter_adaptive_scaffold/CHANGELOG.md
Comment (Renzo-Olivares) [inline packages/flutter_adaptive_scaffold/CHANGELOG.md:4]: nit: period at the end of this line.

**Review by Renzo-Olivares — APPROVED** [review]
I added a few fly-by nits. Still going over the changes to `pickWidget`.

**Comment by martijn00** [comment]
@Renzo-Olivares fixed those now.

File: packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart
Comment (gspencergoog) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart]: Please check all AI-generated code carefully before using it. They are the masters of introducing subtle bugs.

Comment (diegotori) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart]: @gspencergoog so far, that code is simple enough that I don't foresee any bugs, based on how it's written.

In other words, given a collection of Breakpoints to configs, when the current breakpoint is active:

1. If there's a platform for it, immediately return it.
2. If it has a valid config, save it as our chosen config
3. If the next active breakpoint has a valid config, then override the currently chosen config
4. However, if the next active breakpoint has no config, then clear out our currently chosen config.
5. Repeat steps 1 through 4.

Technically, this now accounts for cases where you set your slot's builder to `AdaptiveScaffold.emptyBuilder`, so that it clears out the chosen config in that algorithm. That way, it respects the user's choice in terms of whether or not they want to build a widget for a given slot.

File: packages/flutter_adaptive_scaffold/test/breakpoint_test.dart
Comment (Renzo-Olivares) [inline packages/flutter_adaptive_scaffold/test/breakpoint_test.dart:215]: nit: period at the end of this line.

File: packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart
Comment (diegotori) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart]: @martijn00, @gspencergoog might have a point. If the current AI derived solution is too risky, then it might make sense to fall back to the more naive approach as it was previously written:

```dart
  /// Given a context and a config, it returns the [SlotLayoutConfig] that will 
  /// be chosen from the config under the context's conditions.
  static SlotLayoutConfig? pickWidget(
      BuildContext context, Map<Breakpoint, SlotLayoutConfig?> config) {
    SlotLayoutConfig? chosenWidget;
    SlotLayoutConfig? fallbackWidget;
    config.forEach((Breakpoint breakpoint, SlotLayoutConfig? pickedWidget) {
      if (breakpoint.isActive(context)) {
        if (breakpoint.platform != null) {
          chosenWidget = pickedWidget;
        } else if (chosenWidget == null && pickedWidget != null) {
          fallbackWidget = pickedWidget;
        } else if (chosenWidget == null && fallbackWidget != null && pickedWidget == null) {
          // We previously had a fallback widget, but our slot has no widget.
          // That means don't display anything for it.
          fallbackWidget = null;
        }
      }
    });
    chosenWidget ??= fallbackWidget;
    return chosenWidget;
  }
```

@gspencergoog please let me know if the above is safer than what Gemini produced. Thanks.

Comment (Renzo-Olivares) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart]: Would this be simpler?

```dart
for (final Breakpoint breakpoint in config.keys) {
  if (breakpoint.isActive(context)) {
    final SlotLayoutConfig? pickedWidget = config[breakpoint];
    if (breakpoint.platform != null && pickedWidget != null) {
      return pickedWidget;
    } else {
      chosenWidget = pickedWidget;
    }
  }
}
```

if the breakpoint has a platform then prefer that, if not then we keep overriding the chosenWidget, whether that be null or not.

**Review by gspencergoog — CHANGES_REQUESTED** [review]
Sorry, I need to put this PR on hold for a bit while I check on our policies about AI-generated code.  Hopefully it won't affect things.

File: packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart
Comment (diegotori) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart]: @gspencergoog if AI generated code is an issue, here's an optimized version, combining parts of @Renzo-Olivares's changes with @martijn00's original ones:

```dart
  /// Given a context and a config, it returns the [SlotLayoutConfig] that will
  /// be chosen from the config under the context's conditions.
  static SlotLayoutConfig? pickWidget(
      BuildContext context, Map<Breakpoint, SlotLayoutConfig?> config) {
    SlotLayoutConfig? chosenWidget;
    SlotLayoutConfig? fallbackWidget;
    config.forEach((Breakpoint breakpoint, SlotLayoutConfig? pickedWidget) {
      if (breakpoint.isActive(context)) {
        if (breakpoint.platform != null && pickedWidget != null) {
          chosenWidget = pickedWidget;
        } else {
          fallbackWidget = pickedWidget;
        }
      }
    });
    chosenWidget ??= fallbackWidget;
    return chosenWidget;
  }
```

**Review by gspencergoog — APPROVED** [review]
Ok, all good to go.

(Sorry, policy has changed a lot over the last year, and I wanted to make sure we could still accept it).

**Comment by martijn00** [comment]
@Renzo-Olivares fixed the dots.

File: packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart
Comment (Renzo-Olivares) [reply packages/flutter_adaptive_scaffold/lib/src/slot_layout.dart]: LGTM considering it has less levels of branching.

**Pull request merged**
